File: src/commands/get.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Get command module for the ctag CLI tool.

This module defines the 'get' command for retrieving and displaying tags
from Confluence pages matching a CQL expression.
"""

import json
from typing import Dict, List, Optional, Set

import click

from src.cql import CQLProcessor
from src.interactive import InteractiveHandler
from src.models.search_results import SearchResultItem
from src.tags import TagManager
from src.utils import sanitize_text


@click.command()
@click.argument("cql_expression")
@click.option(
    "--format",
    "output_format",
    type=click.Choice(["table", "json"], case_sensitive=False),
    default="table",
    help="Output format",
)
@click.option(
    "--show-pages/--no-show-pages",
    default=True,
    help="Include page titles and spaces in output",
)
@click.option("--tags-only", is_flag=True, help="Show only unique tags across all pages")
@click.option("--interactive", is_flag=True, help="Browse results interactively")
@click.option(
    "--abort-key",
    default="q",
    help="Key to abort all operations in interactive mode",
)
@click.option(
    "--cql-exclude",
    required=False,
    help="CQL expression to match pages that should be excluded",
)
@click.option("--output-file", type=click.Path(), help="Save results to file")
@click.pass_context
def get(
    ctx,
    cql_expression,
    output_format,
    show_pages,
    tags_only,
    interactive,
    abort_key,
    cql_exclude,
    output_file,
):
    """
    Get tags from pages matching CQL expression.

    CQL_EXPRESSION is a Confluence Query Language expression.

    Examples:
        ctag get "space = DOCS"
        ctag get "space = DOCS" --tags-only
        ctag get "lastmodified > -7d" --format json
        ctag get "title ~ 'Project*'" --interactive
    """
    confluence = ctx.obj["CONFLUENCE"]

    # Initialize our processors
    cql_processor = CQLProcessor(confluence)
    tag_manager = TagManager(confluence)

    # Get matching pages
    click.echo(f"Finding pages matching: {cql_expression}")
    pages = cql_processor.get_all_results(cql_expression)

    if not pages:
        click.echo("No pages found matching the CQL expression.")
        return

    click.echo(f"Found {len(pages)} matching pages.")

    # Apply exclusion if specified
    if cql_exclude:
        click.echo(f"Finding pages to exclude: {cql_exclude}")
        excluded_pages = cql_processor.get_all_results(cql_exclude)
        if excluded_pages:
            original_count = len(pages)
            pages = filter_excluded_pages(pages, excluded_pages)
            click.echo(
                f"Excluded {
        original_count -
        len(pages)} pages. {
            len(pages)} pages remaining."
            )

    # Collect page data with tags
    page_data = []
    all_tags = set()

    click.echo("Retrieving tags for pages...")

    for page in pages:
        # Get the page ID from the content object or dictionary
        page_id = None
        if page.content:
            if isinstance(page.content, dict):
                page_id = page.content.get("id")
            elif hasattr(page.content, "id"):
                page_id = page.content.id

        page_title = sanitize_text(page.title if page.title else "Unknown")

        # Get space information if available
        page_space = "Unknown"
        if hasattr(page, "resultGlobalContainer"):
            if isinstance(page.resultGlobalContainer, dict):
                page_space = page.resultGlobalContainer.get("title", "Unknown")
            elif hasattr(page.resultGlobalContainer, "title"):
                page_space = page.resultGlobalContainer.title

        if not page_id:
            click.echo(f"Warning: Skipping page with no ID: {page_title}", err=True)
            continue

        # Get tags for this page
        page_tags = tag_manager.get_page_tags(page_id)
        all_tags.update(page_tags)

        page_info = {
            "id": page_id,
            "title": page_title,
            "space": page_space,
            "tags": page_tags,
        }
        page_data.append(page_info)

    # Handle interactive mode
    if interactive:
        interactive_handler = InteractiveHandler(default_response=True, abort_value=abort_key)
        page_data = handle_interactive_browsing(page_data, interactive_handler)

    # Generate output based on format and options
    if tags_only:
        output_content = format_tags_only_output(all_tags, output_format)
    else:
        output_content = format_page_data_output(page_data, output_format, show_pages)

    # Output results
    if output_file:
        with open(output_file, "w", encoding="utf-8") as f:
            f.write(output_content)
        click.echo(f"Results saved to {output_file}")
    else:
        click.echo(output_content)

    # Display summary
    click.echo(f"\nSummary:")
    click.echo(f"  Total pages processed: {len(page_data)}")
    click.echo(f"  Unique tags found: {len(all_tags)}")


def filter_excluded_pages(
    pages: List[SearchResultItem], excluded_pages: List[SearchResultItem]
) -> List[SearchResultItem]:
    """Filter out pages that are in the excluded_pages list based on page ID.

    Args:
        pages: List of pages to filter
        excluded_pages: List of pages to exclude

    Returns:
        Filtered list of pages
    """
    excluded_ids = []
    for page in excluded_pages:
        if page.content:
            if isinstance(page.content, dict):
                page_id = page.content.get("id")
                if page_id:
                    excluded_ids.append(page_id)
            elif hasattr(page.content, "id"):
                if page.content.id:
                    excluded_ids.append(page.content.id)

    filtered_pages = []
    for page in pages:
        include = True
        if page.content:
            if isinstance(page.content, dict):
                page_id = page.content.get("id")
                if page_id and page_id in excluded_ids:
                    include = False
            elif hasattr(page.content, "id"):
                if page.content.id and page.content.id in excluded_ids:
                    include = False

        if include:
            filtered_pages.append(page)

    return filtered_pages


def handle_interactive_browsing(page_data: List[Dict], interactive_handler) -> List[Dict]:
    """Handle interactive browsing of page results.

    Args:
        page_data: List of page dictionaries
        interactive_handler: Handler for interactive confirmations

    Returns:
        Filtered list of pages based on user selections
    """
    selected_pages = []

    for page in page_data:
        page_info = f"'{
    page['title']}' (Space: {
        page['space']}, Tags: {
            ', '.join(
                page['tags']) if page['tags'] else 'None'})"
        if interactive_handler.confirm_action(page_info, "Include in results"):
            selected_pages.append(page)

    return selected_pages


def format_tags_only_output(tags: Set[str], output_format: str) -> str:
    """Format output showing only unique tags.

    Args:
        tags: Set of unique tags
        output_format: Output format ('table', 'json')

    Returns:
        Formatted output string
    """
    sorted_tags = sorted(tags)

    if output_format == "json":
        return json.dumps(sorted_tags, indent=2, ensure_ascii=False)
    else:  # table format
        if not sorted_tags:
            return "No tags found."

        output_lines = ["Tags found:"]
        output_lines.append("=" * 50)
        for tag in sorted_tags:
            output_lines.append(f"  {tag}")
        return "\n".join(output_lines)


def format_page_data_output(page_data: List[Dict], output_format: str, show_pages: bool) -> str:
    """Format output showing page data with tags.

    Args:
        page_data: List of page dictionaries
        output_format: Output format ('table', 'json')
        show_pages: Whether to include page information

    Returns:
        Formatted output string
    """
    if output_format == "json":
        if show_pages:
            return json.dumps(page_data, indent=2, ensure_ascii=False)
        else:
            # Extract just tags
            all_tags = []
            for page in page_data:
                all_tags.extend(page["tags"])
            return json.dumps(sorted(set(all_tags)), indent=2, ensure_ascii=False)

    else:  # table format
        if not page_data:
            return "No pages found."

        output_lines = []

        if show_pages:
            # Calculate column widths
            max_title_len = max(len(page["title"]) for page in page_data)
            max_space_len = max(len(page["space"]) for page in page_data)
            title_width = min(max_title_len, 50)  # Cap at 50 chars
            space_width = min(max_space_len, 20)  # Cap at 20 chars

            # Header
            header = f"{'Title':<{title_width}} {'Space':<{space_width}} Tags"
            output_lines.append(header)
            output_lines.append("=" * len(header))

            # Data rows
            for page in page_data:
                title = page["title"][:title_width]
                space = page["space"][:space_width]
                tags = ", ".join(page["tags"]) if page["tags"] else "(no tags)"

                row = f"{title:<{title_width}} {space:<{space_width}} {tags}"
                output_lines.append(row)
        else:
            # Show only unique tags
            all_tags = set()
            for page in page_data:
                all_tags.update(page["tags"])

            if all_tags:
                output_lines.append("Tags found:")
                output_lines.append("=" * 50)
                for tag in sorted(all_tags):
                    output_lines.append(f"  {tag}")
            else:
                output_lines.append("No tags found.")

        return "\n".join(output_lines)

File: src/commands/get.rs
use crate::api::{filter_excluded_pages, sanitize_text, ConfluenceClient};
use crate::models::OutputFormat;
use crate::ui;
use anyhow::Result;
use clap::Args;
use comfy_table::modifiers::UTF8_ROUND_CORNERS;
use comfy_table::presets::UTF8_FULL;
use comfy_table::{Attribute, Cell, Color, ContentArrangement, Table};
use serde::Serialize;
use std::collections::HashSet;
use terminal_size::{terminal_size, Width};

#[derive(Args)]
pub struct GetArgs {
    /// CQL expression to match pages
    pub cql_expression: String,

    /// Include page titles and spaces in output
    #[arg(long, default_value_t = true)]
    pub show_pages: bool,

    /// Show only unique tags across all pages
    #[arg(long)]
    pub tags_only: bool,

    /// Browse results interactively
    #[arg(long)]
    pub interactive: bool,

    /// Key to abort all operations in interactive mode
    #[arg(long, default_value = "q")]
    pub abort_key: String,

    /// CQL expression to match pages that should be excluded
    #[arg(long)]
    pub cql_exclude: Option<String>,

    /// Save results to file
    #[arg(long)]
    pub output_file: Option<String>,
}

#[derive(Serialize)]
struct PageData {
    id: String,
    title: String,
    space: String,
    tags: Vec<String>,
}

pub fn run(
    args: GetArgs,
    client: &ConfluenceClient,
    _dry_run: bool,
    show_progress: bool,
    format: OutputFormat,
) -> Result<()> {
    let verbose = format == OutputFormat::Verbose;
    let is_structured = format == OutputFormat::Json || format == OutputFormat::Csv;

    if verbose {
        ui::print_header("GET TAGS");
    }

    // Get matching pages
    let spinner = if (verbose || !show_progress) && !is_structured {
        Some(ui::create_spinner(&format!(
            "Finding pages matching: {}",
            args.cql_expression
        )))
    } else {
        None
    };
    let mut pages = client.get_all_cql_results(&args.cql_expression, 100)?;
    if let Some(s) = &spinner {
        s.finish_and_clear();
    }

    if pages.is_empty() {
        match format {
            OutputFormat::Json => println!("[]"),
            OutputFormat::Csv => println!(), // Empty CSV
            _ => ui::print_warning("No pages found matching the CQL expression."),
        }
        return Ok(());
    }
    if verbose {
        ui::print_info(&format!("Found {} matching pages.", pages.len()));
    }

    // Apply exclusion if specified
    if let Some(cql_exclude) = &args.cql_exclude {
        let spinner = if (verbose || !show_progress) && !is_structured {
            Some(ui::create_spinner(&format!(
                "Finding pages to exclude: {}",
                cql_exclude
            )))
        } else {
            None
        };
        let excluded_pages = client.get_all_cql_results(cql_exclude, 100)?;
        if let Some(s) = &spinner {
            s.finish_and_clear();
        }
        if !excluded_pages.is_empty() {
            let original_count = pages.len();
            pages = filter_excluded_pages(pages, &excluded_pages);
            if verbose {
                ui::print_info(&format!(
                    "Excluded {} pages. {} pages remaining.",
                    original_count - pages.len(),
                    pages.len()
                ));
            }
        }
    }

    // Collect page data with tags
    if verbose {
        ui::print_step("Retrieving tags for pages...");
    }
    let mut page_data = Vec::new();
    let mut all_tags = HashSet::new();

    let progress = if show_progress && !is_structured {
        Some(ui::create_progress_bar(pages.len() as u64))
    } else {
        None
    };

    for page in &pages {
        let page_id = match &page.content {
            Some(content) => match &content.id {
                Some(id) => id.clone(),
                None => {
                    ui::print_warning("Skipping page with no ID");
                    continue;
                }
            },
            None => {
                ui::print_warning("Skipping page with no content");
                continue;
            }
        };
        let title = sanitize_text(page.title.as_deref().unwrap_or("Unknown"));
        let space = page
            .content
            .as_ref()
            .and_then(|c| c.space.as_ref())
            .and_then(|s| s.name.as_deref())
            .or_else(|| page.space.as_ref().and_then(|s| s.name.as_deref())) // Added check for page.space
            .or_else(|| {
                page.result_global_container
                    .as_ref()
                    .and_then(|c| c.title.as_deref())
            })
            .unwrap_or("Unknown")
            .to_string();

        let tags = client.get_page_tags(&page_id).unwrap_or_default();
        all_tags.extend(tags.iter().cloned());
        page_data.push(PageData {
            id: page_id,
            title,
            space,
            tags,
        });

        if let Some(p) = &progress {
            p.inc(1);
        }
    }

    if let Some(p) = &progress {
        p.finish_and_clear();
    }

    // Generate output
    let output_content = if args.tags_only {
        format_tags_only(&all_tags, &format)
    } else {
        format_page_data(&page_data, &format, args.show_pages, client.base_url())
    };

    // Output results
    if let Some(file_path) = args.output_file {
        std::fs::write(&file_path, output_content)?;
        if verbose {
            ui::print_success(&format!("Results saved to {}", file_path));
        }
    } else {
        println!("{}", output_content);
    }

    if verbose {
        eprintln!();
        ui::print_info(&format!("Total pages processed: {}", page_data.len()));
        ui::print_info(&format!("Unique tags found: {}", all_tags.len()));
    }

    Ok(())
}

fn format_tags_only(tags: &HashSet<String>, format: &OutputFormat) -> String {
    let mut sorted_tags: Vec<_> = tags.iter().collect();
    sorted_tags.sort();
    match format {
        OutputFormat::Json => serde_json::to_string_pretty(&sorted_tags).unwrap_or_default(),
        OutputFormat::Csv => {
            let mut wtr = csv::Writer::from_writer(vec![]);
            #[derive(Serialize)]
            struct TagCsv<'a> {
                tag: &'a str,
            }
            for tag in sorted_tags {
                wtr.serialize(TagCsv { tag }).unwrap();
            }
            String::from_utf8(wtr.into_inner().unwrap()).unwrap()
        }
        OutputFormat::Simple | OutputFormat::Verbose => {
            if sorted_tags.is_empty() {
                return "No tags found.".to_string();
            }
            let mut table = Table::new();
            table
                .load_preset(UTF8_FULL)
                .apply_modifier(UTF8_ROUND_CORNERS)
                .set_header(vec![Cell::new("Tag")
                    .add_attribute(Attribute::Bold)
                    .fg(Color::Cyan)]);

            for tag in sorted_tags {
                table.add_row(vec![tag]);
            }
            table.to_string()
        }
    }
}

fn format_page_data(
    page_data: &[PageData],
    format: &OutputFormat,
    show_pages: bool,
    base_url: &str,
) -> String {
    match format {
        OutputFormat::Json => {
            if show_pages {
                serde_json::to_string_pretty(&page_data).unwrap_or_default()
            } else {
                let mut all_tags: HashSet<String> = HashSet::new();
                for page in page_data {
                    all_tags.extend(page.tags.iter().cloned());
                }
                let mut sorted: Vec<_> = all_tags.into_iter().collect();
                sorted.sort();
                serde_json::to_string_pretty(&sorted).unwrap_or_default()
            }
        }
        OutputFormat::Csv => {
            let mut wtr = csv::Writer::from_writer(vec![]);
            if show_pages {
                #[derive(Serialize)]
                struct PageDataCsv<'a> {
                    id: &'a str,
                    title: &'a str,
                    space: &'a str,
                    tags: String,
                }

                for page in page_data {
                    wtr.serialize(PageDataCsv {
                        id: &page.id,
                        title: &page.title,
                        space: &page.space,
                        tags: page.tags.join(", "),
                    })
                    .unwrap();
                }
            } else {
                let mut all_tags: HashSet<String> = HashSet::new();
                for page in page_data {
                    all_tags.extend(page.tags.iter().cloned());
                }
                let mut sorted: Vec<_> = all_tags.into_iter().collect();
                sorted.sort();

                #[derive(Serialize)]
                struct TagCsv<'a> {
                    tag: &'a str,
                }

                for tag in sorted {
                    wtr.serialize(TagCsv { tag: &tag }).unwrap();
                }
            }
            String::from_utf8(wtr.into_inner().unwrap()).unwrap()
        }
        OutputFormat::Simple | OutputFormat::Verbose => {
            if page_data.is_empty() {
                return "No pages found.".to_string();
            }
            if show_pages {
                let mut table = Table::new();
                table
                    .load_preset(UTF8_FULL)
                    .apply_modifier(UTF8_ROUND_CORNERS)
                    .set_content_arrangement(ContentArrangement::Dynamic);

                let width = terminal_size().map(|(Width(w), _)| w).unwrap_or(120);
                table.set_width(width.saturating_sub(4));

                table.set_header(vec![
                    Cell::new("Title")
                        .add_attribute(Attribute::Bold)
                        .fg(Color::Cyan),
                    Cell::new("Space")
                        .add_attribute(Attribute::Bold)
                        .fg(Color::Cyan),
                    Cell::new("Tags")
                        .add_attribute(Attribute::Bold)
                        .fg(Color::Cyan),
                ]);

                use comfy_table::{ColumnConstraint::*, Width::*};
                table
                    .column_mut(0)
                    .unwrap()
                    .set_constraint(LowerBoundary(Fixed(40)));
                table
                    .column_mut(1)
                    .unwrap()
                    .set_constraint(LowerBoundary(Fixed(15)));
                table
                    .column_mut(2)
                    .unwrap()
                    .set_constraint(LowerBoundary(Fixed(20)));

                for page in page_data {
                    let tags = if page.tags.is_empty() {
                        "-".to_string()
                    } else {
                        page.tags.join(", ")
                    };

                    let title_content = if format == &OutputFormat::Verbose {
                        // Create a terminal hyperlink if in verbose mode
                        format!(
                            "\x1b]8;;{}/wiki/pages/viewpage.action?pageId={}\x1b\\{}\x1b]8;;\x1b\\",
                            base_url, page.id, page.title
                        )
                    } else {
                        page.title.clone()
                    };

                    table.add_row(vec![
                        Cell::new(title_content),
                        Cell::new(&page.space),
                        Cell::new(tags),
                    ]);
                }
                table.to_string()
            } else {
                let mut all_tags: HashSet<String> = HashSet::new();
                for page in page_data {
                    all_tags.extend(page.tags.iter().cloned());
                }

                if all_tags.is_empty() {
                    "No tags found.".to_string()
                } else {
                    let mut table = Table::new();
                    table
                        .load_preset(UTF8_FULL)
                        .apply_modifier(UTF8_ROUND_CORNERS)
                        .set_header(vec![Cell::new("Tag")
                            .add_attribute(Attribute::Bold)
                            .fg(Color::Cyan)]);

                    let mut sorted: Vec<_> = all_tags.into_iter().collect();
                    sorted.sort();
                    for tag in sorted {
                        table.add_row(vec![tag]);
                    }
                    table.to_string()
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashSet;

    #[test]
    fn format_tags_only_table_empty() {
        let tags: HashSet<String> = HashSet::new();
        let out = format_tags_only(&tags, &OutputFormat::Simple);
        assert_eq!(out.trim(), "No tags found.");
    }

    #[test]
    fn format_tags_only_json_sorted() {
        let mut tags: HashSet<String> = HashSet::new();
        tags.insert("b".to_string());
        tags.insert("a".to_string());
        let out = format_tags_only(&tags, &OutputFormat::Json);
        let parsed: Vec<String> = serde_json::from_str(&out).unwrap();
        assert_eq!(parsed, vec!["a".to_string(), "b".to_string()]);
    }
}
